### ISSUE TITLE
Different Regexp behavior from MRI

### ISSUE THREAD

**Issue by nirvdrum** [jruby/jruby#2905]
It looks like somewhere between 1.9.3 and 2.2.2, MRI changed its behavior around overriding `String#=~`.  JRuby 9.0.0.0.pre2 special-cases this behavior in:

https://github.com/jruby/jruby/blob/9.0.0.0.pre2/core/src/main/java/org/jruby/ir/runtime/IRRuntimeHelpers.java#L577-L579

It looks like the correct behavior is to always do a dispatch to `=~` to pick up any overrides.

E.g.,

MRI 2.2.2:

```
> ruby -v -e 'class String; def =~(regexp); :weird_value; end; end; p "a" =~ /./'
-e:1: warning: method redefined; discarding old =~
:weird_value
```

JRuby 9.0.0.0.pre2:

```
> bin/jruby -v -e 'class String; def =~(regexp); :weird_value; end; end; p "a" =~ /./'
jruby 9.0.0.0-SNAPSHOT (2.2.2) 2015-04-30 ef1fc91 Java HotSpot(TM) 64-Bit Server VM 25.45-b02 on 1.8.0_45-b14 +jit [linux-amd64]
0
```

**Comment by nirvdrum** [comment]
Interesting, MRI 1.9.3 prints the same warning, but then prints the match index.

**Comment by headius** [comment]
Might just be simplest to make all matches dyncall; caching and indy will make up the difference between this and the direct call.

**Comment by headius** [comment]
Oops, forgot to tag the commit. This should be fixed on master in 8d67e58.
